ground/gcs/src/plugins/config/configautotunewidget.cpp

```diff
@@ -395,417 +333,162 @@ void ConfigAutotuneWidget::stuffShareForm(AutotuneFinalPage *autotuneShareForm)
     }
 }
 
-// XXX appropriately encapsulate all this crap
-#define ATFLASH_MAGIC 0x656e755480008041 /* A...Tune */
-struct at_flash_header
+void ConfigAutotuneWidget::openAutotuneFile()
 {
-    uint64_t magic;
-    uint16_t wiggle_points;
-    uint16_t aux_data_len;
-    uint16_t sample_rate;
+    QString fileName =
+        QFileDialog::getOpenFileName(this, tr("Open autotune partition"), "",
+                                     tr("Partition image Files (*.bin) ;; All files (*.*)"));
 
-    // Consider total number of averages here
-    uint16_t resv;
-};
+    if (fileName.isEmpty()) {
+        return;
+    }
 
-struct at_measurement
-{
-    float y[3]; /* Gyro measurements */
-    float u[3]; /* Actuator desired */
-};
+    QFile file(fileName);
+    if (!file.open(QIODevice::ReadOnly)) {
+        /* XXX error dialog */
+        return;
+    }
 
-struct at_flash
-{
-    struct at_flash_header hdr;
+    AutotunedValues vals;
+    memset(&vals, 0, sizeof(vals));
+    vals.data = new QByteArray(file.readAll());
 
-    struct at_measurement data[];
-};
+    openAutotuneDialog(false, &vals);
+}
 
-/* Run a Butterworth biquad filter on a circular buffer.  First go around once
- * to "prime" the filter, then actually filter in place.  This is derived from
- * @glowtape's excellent flight implementation
- */
-void biquad_filter_in_place(float cutoff, int pts, float *data)
+void ConfigAutotuneWidget::openAutotuneDialog()
 {
-    float f = 1.0f / tan(M_PI * cutoff);
-    float q = 1.4142f;
-
-    float y2 = 0, y1 = 0, x2 = 0, x1 = 0;
+    openAutotuneDialog(false);
+}
 
-    float b0 = 1.0f / (1.0f + q * f + f * f);
-    float a1 = 2.0f * (f * f - 1.0f) * b0;
-    float a2 = -(1.0f - q * f + f * f) * b0;
+void ConfigAutotuneWidget::openAutotuneDialog(bool autoOpened,
+        AutotunedValues *precalc_vals)
+{
+    QWizard wizard(NULL, Qt::Dialog | Qt::CustomizeWindowHint | Qt::WindowTitleHint
+                       | Qt::WindowCloseButtonHint);
 
-    for (int i = 0; i < pts; i++) {
-        float y = b0 * (data[i] + 2.0f * x1 + x2) + a1 * y1 + a2 * y2;
+    wizard.setPixmap(QWizard::BackgroundPixmap, QPixmap(":/configgadget/images/autotunebg.png"));
 
-        y2 = y1;
-        y1 = y;
+    // Used to track state across the entire wizard path.
+    AutotunedValues av;
```

Comment:
Shadows the member variable? I would strongly prefer the member variable to have a better name since it has a _long_ lifetime and large scope, so probably best to change it to `tuneValues` or something.